### PR TITLE
[FEAT] Prepare environment list for export and normal assignment

### PR DESCRIPTION
**Pull request by itislu**
- [x] First merge #163 
---
- Add export flag to env struct
- Handle the following default environment variables correctly:
  - If no `OLDPWD` key exists, create one with no value.
  - `PWD` should always be set with `getcwd()` by current shell at startup.
  - If `OLDPWD` exists and its value is not a real directory, delete `OLDPWD` entirely (permissions don't matter).

### CODE REVIEW

**Comment by itislu** [comment]
It's very hard to reproduce what's wrong in the tester.
I will try to implement printouts of what the tester compares with bash.

File: source/utils/final_cmd_table_utils.c
Comment (itislu) [inline source/utils/final_cmd_table_utils.c:1]: Only put env variables that are both exported and have a value that is non-null (empty string is fine) into the envp array.

**Comment by LeaYeh** [comment]
Why you remove env from shell structure

**Comment by itislu** [comment]
> Why you remove env from shell structure

But I didn't 😧
I just renamed it from `envp` to `env` to avoid confusion with the real `envp` that one can get from the third argument of the main function.

**Comment by itislu** [comment]
In case you meant why I removed `char **envp` from the main function's paramters...
I did it bc using `envp` is actually not POSIX compliant, so not guaranteed to be portable.
It is recommended to use the global variable `extern char **environ` defined in `unistd.h` instead.

> In Unix systems you can define main a third way, using three arguments:
> 
> int main (int argc, char *argv[], char *envp[])
> The first two arguments are just the same. The third argument envp gives the program’s environment; it is the same as the value of environ. See [Environment Variables](https://www.gnu.org/software/libc/manual/html_node/Environment-Variables.html). POSIX.1 does not allow this three-argument form, so to be portable it is best to write main to take two arguments, and use the value of environ.

https://www.gnu.org/software/libc/manual/html_node/Program-Arguments.html
https://stackoverflow.com/questions/10321435/is-char-envp-as-a-third-argument-to-main-portable
https://pubs.opengroup.org/onlinepubs/9699919799/basedefs/V1_chap08.html

**Comment by LeaYeh** [comment]
Did you check others comments?

**Comment by itislu** [comment]
Sorry, I don't understand. Others' comments about what?

File: include/builtins.h [hunk 1]
```diff
@@ -23,12 +23,12 @@ typedef enum e_exit_args_error
 	NOT_NUMERIC,
 }	t_exit_args_error;
 
-int		ft_exec_env(t_shell *shell);
-int		ft_exec_echo(char **args);
+int		ft_exec_env(char *env[]);
```
[the remark this answers is not in the document]
Comment (itislu) [reply include/builtins.h:26]: In C, there is no technical difference.
However, I read and saw in others' code to use `char *[]` if the expected parameter will be an array, and `char **` if the expected parameter will be the address of a `char *`.

I think this makes the affected functions easier to use and read.

File: include/defines.h [hunk 2]
```diff
@@ -238,10 +238,17 @@ typedef enum e_expander_op
 	E_HEREDOC		= 0b100
 }	t_expander_op;
 
+typedef enum e_export
```
[the remark this answers is not in the document]
Comment (itislu) [reply include/defines.h:241]: With normal assignment there will probably be more states.
I'm currently working on it and already added a third enum.

File: source/init/setup_env_list.c [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+		return (false);
+	return (true);
+}
+
+/*
+ * PWD should always be set by current shell.
+ * If OLDPWD exists and its value is not a real directory,
+ * delete OLDPWD entirely (permissions don't matter).
+ */
+bool	process_env_str_to_env_list(char *env_str, t_list **env_list)
```
[the remark this answers is not in the document]
Comment (itislu) [reply source/init/setup_env_list.c:24]: * This function will be refactored and split up with a new name (`process_str_to_env_list()`) in #167 
* Commit: 048c6683e23c9560849d95649711e1d9a10780e5

File: include/init.h [hunk 4]
```diff
@@ -0,0 +1,10 @@
+#ifndef INIT_H
```
[the remark this answers is not in the document]
Comment (itislu) [reply include/init.h:1]: What exactly would you like to separate?

File: source/builtins/cd/update_pwd_env.c [hunk 5]
```diff
@@ -11,38 +11,36 @@
 /* ************************************************************************** */
 
 #include "cd.h"
-#include "minishell.h"
+#include "utils.h"
```
[the remark this answers is not in the document]
Comment (itislu) [reply source/builtins/cd/update_pwd_env.c:14]: Ok I will rename it to `pwd_env_update.c`.

**Review by LeaYeh — APPROVED** [review]

**Pull request merged**